### DISCUSSION TITLE
Add option to install MIT (or any other) dependencies only

### DISCUSSION THREAD

**Discussion by MasterKym** [npm/feedback#447] [CLI] [answered]
As many of you know, different license types force different rules.
I was thinking if we can have something like `npm install -license=<some-licence> <package-name>`, where the command would resolve only if the package satisfies the license condition. This way one can be sure that all packages being used in their project align with their intended use.

Thoughts ?

**Comment by ljharb** [comment]
You can use https://npmjs.com/licensee in a postinstall script to validate the licenses of your dependencies, and it will fail the overarching install if they're invalid. You can also run it separately in CI.

**Reply by MasterKym** [reply]
I know I can do that, but I was thinking this might be suitable for a flag of the install command.

**Comment by MylesBorins** [comment, marked as answer]
@MasterKym I've though about something like this making a bit more sense project wide as a compliance check similar to audit. There is an existing RFc for this feature!

https://github.com/npm/rfcs/pull/182

**Reply by ljharb** [reply]
That would be like licensee; the OP seems to be asking for a non-matching license to never be installed in the first place.

**Reply by MasterKym** [reply]
@MylesBorins Thank you for getting back.
As @ljharb said, I am actually looking to see if we can have something to only allow packages with certain licenses to be installed. Installing then auditing seems like a bit of an extra step ( still necessary for serious projects, don't get me wrong.. but an extra step ).

**Reply by MylesBorins** [reply]
My thinking here is that these are entirely related. If we have a way to audit the tree we will also need a way to define a policy as to how the audit is performed. Once you have a policy it is a fairly straight forward next step to enforce said policy on install.

Currently you can cause `npm audit` to exit with a non-zero code by passing `--audit-level` and providing a min-bar of what is necessary for a 0 exit code. It doesn't appear that you can pass this flag to install, but that seems like a reasonable feature to explore independently... e.g. install should fail if audit fails.

If we have a combination of Audit licenses, a way to define policy for licenses, and a way to force npm install to fail on audit, that would cover the exact use case you are asking for using primitives that imho are much more flexible across a variety of things we may want to audit (security, license, system-access).

Might be good to expand the RFC for auditing licenses to make sure it has policies + `--audit-level` in scope and might also be worth an additional RFC on allowing `--audit-level` to be passed to install

**Reply by MasterKym** [reply]
> My thinking here is that these are entirely related. If we have a way to audit the tree we will also need a way to define a policy as to how the audit is performed. Once you have a policy it is a fairly straight forward next step to enforce said policy on install.

You're right about this.
> Currently you can cause `npm audit` to exit with a non-zero code by passing `--audit-level` and providing a min-bar of what is necessary for a 0 exit code. It doesn't appear that you can pass this flag to install, but that seems like a reasonable feature to explore independently... e.g. install should fail if audit fails.

Indeed.

> If we have a combination of Audit licenses, a way to define policy for licenses, and a way to force npm install to fail on audit, that would cover the exact use case you are asking for using primitives that imho are much more flexible across a variety of things we may want to audit (security, license, system-access).

Could you further elaborate what you mean by "Audit licenses" ?
I assume for defining a policy for licenses, it is more than just defining an object inside `package.json` that the install checks before proceeding, it would also involve forcing providing a license to upload to npm, include some way to retrieve the license data before doing the install and most likely other things..
Also, you said:
>a way to force npm install to fail on audit

I assume that's not trivial given the way you brought it up. Sorry I am not familiar with the source code of npm.

> Might be good to expand the RFC for auditing licenses to make sure it has policies + `--audit-level` in scope and might also be worth an additional RFC on allowing `--audit-level` to be passed to install

I am mostly concerned about code distribution here (I am aware that if we end up building it it'll have to cover more than just what concerns me, but this is a starting point), so the way the command is built can have a couple of flags, among which one is okay for code distribution.
